x-pack/plugins/observability_solution/investigate/public/plugin.tsx

```diff
@@ -28,18 +31,49 @@ export class InvestigatePlugin
 
   widgetRegistry: WidgetRegistry = new WidgetRegistry();
 
+  registrationPromises: Array<Promise<void>> = [];
+
   constructor(context: PluginInitializerContext<ConfigSchema>) {
     this.logger = context.logger.get();
   }
   setup(coreSetup: CoreSetup, pluginsSetup: InvestigateSetupDependencies): InvestigatePublicSetup {
     return {
-      registerWidget: this.widgetRegistry.registerWidget,
+      register: (callback) => {
+        const registrationPromise = Promise.race([
+          callback(this.widgetRegistry.registerWidget),
+          new Promise<void>((resolve, reject) => {
+            setTimeout(() => {
+              reject(new Error('Timed out running registration function'));
+            }, 30000);
+          }),
+        ]).catch((error) => {
+          this.logger.error(
+            new Error('Encountered an error during widget registration', { cause: error })
+          );
+          return Promise.resolve();
+        });
+
+        this.registrationPromises.push(registrationPromise);
+      },
     };
```

Comment:
There is an opportunity here for a race condition: `register` is asynchronous, but `getWidgetDefinitions` is not. In practice you will probably never run into it, but it might become an issue for tests. You could consider making an Observable out of `getWidgetDefinitions` but there are some implications for code downstream.